src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +474,164 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
-                    " The limit on off-heap memory size is given by the -XX:MaxDirectMemorySize command (or equivalent).");
-          } else {
-            LOGGER.debug("Allocated {}B in {}B chunks.", new Object[]{DebuggingUtils.toBase2SuffixedString(allocated - lastFailureAt), DebuggingUtils.toBase2SuffixedString(currentChunkSize)});
-            currentChunkSize >>>= 1;
-            lastFailureAt = allocated;
-          }
-        } else {
-          buffers.add(b);
-          allocated += b.capacity();
-          
-          if (allocatorLog != null) {
-            allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start, blockDuration, b.capacity(), allocated, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
-          }
-          if (allocated > nextProgressLogAt) {
-            LOGGER.info("Allocation {}% complete", (100 * allocated) / max);
-            nextProgressLogAt += progressStep;
+        if (allocatorLog != null) {
+          allocatorLog.printf("timestamp,threadid,duration,size,physfree,totalswap,freeswap,committed%n");
+        }
+
+        ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+
+        List<Future<Collection<ByteBuffer>>> futures = new ArrayList<Future<Collection<ByteBuffer>>>((int)(toAllocate / maxChunk + 1));
+
+        for (long dispatched = 0; dispatched < toAllocate; ) {
+          final int currentChunkSize = (int)Math.min(maxChunk, toAllocate - dispatched);
+          futures.add(executorService.submit(new Callable<Collection<ByteBuffer>>() {
+            @Override
+            public Collection<ByteBuffer> call() throws Exception {
+              return bufferAllocation(source, currentChunkSize, minChunk, fixed, allocatorLog, start);
+            }
+          }));
+          dispatched += currentChunkSize;
+        }
+
+        executorService.shutdown();
+
+        long allocated = 0;
+        long progressStep = Math.max(PROGRESS_LOGGING_THRESHOLD, (long)(toAllocate * PROGRESS_LOGGING_STEP_SIZE));
+        long nextProgressLogAt = progressStep;
+        AtomicBoolean wasInterrupted = new AtomicBoolean(false);
+
+        for (Future<Collection<ByteBuffer>> future : futures) {
+          Collection<ByteBuffer> result = uninterruptibleGet(future, wasInterrupted);
+          buffers.addAll(result);
+          for(ByteBuffer buffer : result) {
+            allocated += buffer.capacity();
+            if (allocated > nextProgressLogAt) {
+              LOGGER.info("Allocation {}% complete", (100 * allocated) / toAllocate);
+              nextProgressLogAt += progressStep;
+            }
           }
         }
+
+        // Set the interruption flag to tell whoever is interested
+        if(wasInterrupted.get()) {
+          Thread.currentThread().interrupt();
```

Comment:
I would just do the interrupt reasserting inside the uninterruptibleGet method.  It makes the method contract simpler and in the rare case that we do get interrupted it will just result in a single extra loop iteration per allocation.